examples/ip_rtic.rs

```diff
@@ -0,0 +1,222 @@
+
+use log::{Level, LevelFilter, Metadata, Record};
+use smoltcp::iface::{EthernetInterface, EthernetInterfaceBuilder, NeighborCache};
+use smoltcp::socket::{SocketSet, TcpSocket, TcpSocketBuffer};
+use smoltcp::time::Instant;
+use smoltcp::wire::{EthernetAddress, IpAddress, IpCidr, Ipv4Address};
+
+use cortex_m_semihosting::hio::HStdout;
+use rtic::export::DWT;
+use stm32_eth::smoltcp::iface::Neighbor;
+use stm32_eth::smoltcp::socket::{SocketHandle, SocketSetItem};
+use stm32_eth::{Eth, EthPins, PhyAddress, RingEntry};
+
+static mut LOGGER: HioLogger = HioLogger {};
+
+struct HioLogger {}
+
+impl log::Log for HioLogger {
+    fn enabled(&self, metadata: &Metadata) -> bool {
+        metadata.level() <= Level::Trace
+    }
+
+    fn log(&self, record: &Record) {
+        if self.enabled(record.metadata()) {
+            let mut stdout = hio::hstdout().unwrap();
+            writeln!(stdout, "{} - {}", record.level(), record.args()).unwrap();
+        }
+    }
+    fn flush(&self) {}
+}
+
+const MS_PERIOD: u32 = 16_000;
+
+const SRC_MAC: [u8; 6] = [0x00, 0x00, 0xDE, 0xAD, 0xBE, 0xEF];
+
+type Ethernet = EthernetInterface<'static, 'static, 'static, &'static mut Eth<'static, 'static>>;
+type Sockets = SocketSet<'static, 'static, 'static>;
+
+#[rtic::app(device = stm32f4xx_hal::stm32, peripherals = true, monotonic = rtic::cyccnt::CYCCNT)]
+const APP: () = {
+    struct Resources {
+        #[init(0)]
+        time: u64,
+        #[init(false)]
+        eth_pending: bool,
+        server_handle: SocketHandle,
+        sockets: Sockets,
+        interface: Ethernet,
+        stdout: HStdout,
+    }
+
+    #[init(schedule = [ms_tick])]
+    fn init(cx: init::Context) -> init::LateResources {
+        static mut RX_RING: Option<[RingEntry<RxDescriptor>; 8]> = None;
+        static mut TX_RING: Option<[RingEntry<TxDescriptor>; 2]> = None;
+
+        static mut IP_ADDRS: Option<[IpCidr; 1]> = None;
+
+        static mut NEIGHBOR_STORAGE: [Option<(IpAddress, Neighbor)>; 16] = [None; 16];
+
+        static mut ETH: Option<Eth<'static, 'static>> = None;
+
+        static mut SOCKETS_STORAGE: [Option<SocketSetItem<'static, 'static>>; 2] = [None, None];
+        static mut SERVER_RX_BUFFER: [u8; 2048] = [0; 2048];
+        static mut SERVER_TX_BUFFER: [u8; 2048] = [0; 2048];
+
+        unsafe {
+            log::set_logger(&LOGGER).unwrap();
+        }
+        log::set_max_level(LevelFilter::Info);
+
+        let mut stdout = hio::hstdout().unwrap();
+
+        let mut core: rtic::Peripherals = cx.core;
+        let device: stm32f4xx_hal::stm32::Peripherals = cx.device;
+
+        // enable CYCCNT
+        core.DCB.enable_trace();
+        DWT::unlock();
+        core.DWT.enable_cycle_counter();
+
+        let rcc = device.RCC.constrain();
+        // HCLK must be at least 25MHz to use the ethernet peripheral
+        let clocks = rcc.cfgr.sysclk(32.mhz()).hclk(32.mhz()).freeze();
+
+        writeln!(stdout, "Enabling ethernet...").unwrap();
```

Comment:
Why don't use `log` instead of `writeln!(stdout)` ? That way you save some code and a RTIC resource.

Reply:
I used the same thing used in the non-RTIC example for consistency, my code originally used defmt. What exactly do you mean by log? is there a Rust `log` backend for RTT?

Reply:
You already implemented log for semihosting, can't you just call `log::info!` (etc) instead of `writeln!` ?

Reply:
Oh, and about the `defmt` commented out code, maybe you could change it for `log::error!` ?